File: AVCam/Objective-C/AVCam/ItemSelectionViewController.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
The view controller that presents the SSM selection interface.
*/

#include "ItemSelectionViewController.h"


@implementation ItemSelectionViewController
{
	id<ItemSelectionViewControllerDelegate> _delegate;

	NSString *_identifier;
	NSArray<id> *_allItems;
	NSMutableArray<id> *_selectedItems;
	BOOL _allowMultipleSelection;
	NSString *itemCellIdentifier;
}


- (id)initWithDelegate:(id<ItemSelectionViewControllerDelegate>)delegate identifier:(NSString *)identifier allItems:(NSArray*)allitems selectedItems:(NSArray*)selectedItems allowMultipleSelection:(BOOL)allowMultipleSelection
{
	self = [super initWithStyle:UITableViewStyleGrouped];
	_delegate = delegate;
	_identifier = identifier;
	_allItems = allitems;
	_selectedItems = [NSMutableArray arrayWithArray:selectedItems];
	_allowMultipleSelection = allowMultipleSelection;
	itemCellIdentifier = @"item";
	self.tableView.allowsMultipleSelection = allowMultipleSelection;
	
	self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(done:)];
	
	[self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:itemCellIdentifier];
	
	return self;
}

- (IBAction) done:(id)sender
{
	[_delegate itemSelectionViewController:self didFinishSelectingItems:_selectedItems];
	[self dismissViewControllerAnimated:YES completion:nil];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
	id item = _allItems[indexPath.row];
	
	UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:itemCellIdentifier forIndexPath:indexPath];
	cell.tintColor = UIColor.blackColor;
	
	NSString *textLabel = (NSString *)item;
	if ( [textLabel containsString:@"Skin"] != 0) {
		cell.textLabel.text = @"Skin";
	}
	else if ( [textLabel containsString:@"Teeth"] != 0) {
		cell.textLabel.text = @"Teeth";
	}
	else if ( [textLabel containsString:@"Hair"] != 0) {
		cell.textLabel.text = @"Hair";
	}
	else {
		cell.textLabel.text = @"Not supported";
	}
	
	
	if ( [_selectedItems containsObject: item] ) {
		cell.accessoryType = UITableViewCellAccessoryCheckmark;
	}
	else {
		cell.accessoryType = UITableViewCellAccessoryNone;
	}
	
	return cell;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
	return _allItems.count;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	if ( _allowMultipleSelection ) {
		id item = _allItems[indexPath.row];
		
		if ( [_selectedItems containsObject:item] ) {
			[_selectedItems removeObject:item];
		}
		else {
			[_selectedItems addObject:item];
		}
		
		[self.tableView deselectRowAtIndexPath:indexPath animated:true];
		[self.tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
	}
}

@end

File: AVCam/Swift/AVCam/ItemSelectionViewController.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
The view controller that presents the SSM selection interface.
*/

import UIKit
import AVFoundation

protocol ItemSelectionViewControllerDelegate: class {
    func itemSelectionViewController(_ itemSelectionViewController: ItemSelectionViewController,
                                     didFinishSelectingItems selectedItems: [AVSemanticSegmentationMatte.MatteType])
}

class ItemSelectionViewController: UITableViewController {
    
    weak var delegate: ItemSelectionViewControllerDelegate?
    
    let identifier: String
    
    let allItems: [AVSemanticSegmentationMatte.MatteType]
    
    var selectedItems: [AVSemanticSegmentationMatte.MatteType]
    
    let allowsMultipleSelection: Bool
    
    private let itemCellIdentifier = "Item"
    
    init(delegate: ItemSelectionViewControllerDelegate,
         identifier: String,
         allItems: [AVSemanticSegmentationMatte.MatteType],
         selectedItems: [AVSemanticSegmentationMatte.MatteType],
         allowsMultipleSelection: Bool) {
        
        self.delegate = delegate
        self.identifier = identifier
        self.allItems = allItems
        self.selectedItems = selectedItems
        self.allowsMultipleSelection = allowsMultipleSelection
        
        super.init(style: .grouped)
        
        navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: self, action: #selector(done))
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: itemCellIdentifier)
        
        view.tintColor = .black
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("`ItemSelectionViewController` cannot be initialized with `init(coder:)`")
    }
    
    @IBAction private func done() {
        // Notify the delegate that selecting items is finished.
        delegate?.itemSelectionViewController(self, didFinishSelectingItems: selectedItems)
        
        // Dismiss the view controller.
        dismiss(animated: true, completion: nil)
    }
    
    // MARK: UITableViewDataSource
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let ssmType = allItems[indexPath.row]
        
        let cell = tableView.dequeueReusableCell(withIdentifier: itemCellIdentifier, for: indexPath)
        
        // Evaluate the semantic segmentation type to determine the label.
        switch ssmType {
        case .hair:
            cell.textLabel?.text = "Hair"
        case .teeth:
            cell.textLabel?.text = "Teeth"
        case .skin:
            cell.textLabel?.text = "Skin"
        default:
            fatalError("Unknown matte type specified.")
        }
        
        cell.accessoryType = selectedItems.contains(ssmType) ? .checkmark : .none
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return allItems.count
    }
    
    // MARK: - UITableViewDelegate
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if allowsMultipleSelection {
            let item = allItems[indexPath.row]
            
            if selectedItems.contains(item) {
                selectedItems = selectedItems.filter { $0 != item }
            } else {
                selectedItems.append(item)
            }
            
            tableView.deselectRow(at: indexPath, animated: true)
            tableView.reloadRows(at: [indexPath], with: .automatic)
        }
    }
}
